merkle/maphasher/maphasher.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package maphasher provides hashing for maps.
+package maphasher
+
+import (
+	"crypto"
+	"fmt"
+)
+
+// Domain separation prefixes
+const (
+	leafHashPrefix = 0
+	nodeHashPrefix = 1
+)
+
+// Default is a SHA256 based MapHasher for maps.
+var Default = New(crypto.SHA256)
+
+// MapHasher implements a sparse merkel tree hashing algorithm. For testing only.
+// It matches the test vectors generated by other sparse map implememtations,
+// but it does not offer the full N bit security of the underlying hash function.
+type MapHasher struct {
+	crypto.Hash
+	nullHashes [][]byte
+}
+
+// New creates a new merkle.MapHasher using the passed in hash function.
+func New(h crypto.Hash) *MapHasher {
+	m := &MapHasher{Hash: h}
+	m.nullHashes = m.createNullHashes()
+	return m
+}
+
+// String returns a string representation for debugging.
+func (m *MapHasher) String() string {
+	return fmt.Sprintf("MapHasher{%v}", m.Hash)
+}
+
+// HashEmpty returns the hash of an empty branch at a given depth.
+// A depth of 0 indicates the hash of an empty leaf.
+// Empty branches within the tree are plain interior nodes e1 = H(e0, e0) etc.
+func (m *MapHasher) HashEmpty(depth int) []byte {
+	if depth < 0 || depth >= len(m.nullHashes) {
+		panic(fmt.Sprintf("HashEmpty(%v) out of bounds", depth))
+	}
+	return m.nullHashes[depth]
+}
+
+// HashLeaf returns the Merkle tree leaf hash of the data passed in through leaf.
+// The data in leaf is prefixed by the LeafHashPrefix.
```

Comment:
Maybe this would be clearer if you used the same x|y format as the hashed structure in HashChildren below.